metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,278 @@
+import os
+import sys
+import time
+import signal
+import shutil
+import hashlib
+import asyncio
+import tempfile
+from typing import List
+
+
+def hash_command_invocation(command: List[str]):
+    concatenated_string = "".join(command)
+    current_time = str(time.time())
+    concatenated_string += current_time
+    hash_object = hashlib.sha256(concatenated_string.encode())
+    return hash_object.hexdigest()
+
+
+class LogReadTimeoutError(Exception):
+    pass
+
+
+class SubprocessManager(object):
+    def __init__(self):
+        self.commands = {}
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.cleanup()
+
+    async def run_command(self, command: List[str], env=None, cwd=None):
+        command_id = hash_command_invocation(command)
+        self.commands[command_id] = CommandManager(command, env, cwd)
+        await self.commands[command_id].run()
+        return command_id
+
+    def get(self, command_id: str) -> "CommandManager":
+        return self.commands.get(command_id, None)
+
+    async def cleanup(self):
+        for _, v in self.commands.items():
+            await v.cleanup()
+
+
+class CommandManager(object):
+    def __init__(self, command: List[str], env=None, cwd=None):
```

Comment:
Type hints and doc strings for all user-exposed code.

Reply:
done